Adding ng-add schematics

feeloor opened the pull request:
Adding ng-add schematics for Augury which will allow user to install Augury through `ng add @augury/schematics`. 

Schematics will do the following:
1. Add + Install `@augury/core` & `@augury/performance-profiler-plugin` as `devDependencies` to package.json.
2. Add `main.augury.ts` file
3. Add Augury `build` & `serve` configuration to `angular.json`
4. Add Augury `ng serve` script to `package.json`.

Will close #30.

**Known limitations**
I want to default the project name option to the property `defaultProject` in `angular.json` if nothing is entered. Currently there is no way of doing this properly see issue: https://github.com/angular/angular-cli/issues/14067. It would also be great to start of by asking `Want to use "DEFAULT_PROJECT" (Y/n)?` and if no is selected, prompt user to enter the application name.

feeloor committed:
Initial code for ng-add with working build

feeloor committed:
Fixing webpack build to make ng-add work

feeloor committed:
Adding custom param validation

feeloor committed:
Adding tests for schematics

andrewthauer commented on package.json:
```diff
@@ -55,6 +55,9 @@
     "webpack-cli": "^3.1.0",
-    "webpack-filter-warnings-plugin": "^1.2.1"
+    "webpack-filter-warnings-plugin": "^1.2.1",
+    "@angular-devkit/core": "~7.3.7",
```
These seem like they are not in alpha order. Did you add them manually?

feeloor replied:
Yeah, I added it manually and total forgot to order them, I'll update that.

andrewthauer commented on packages/schematics/package.json:
```diff
@@ -0,0 +1,35 @@
+    "build:prod": "cross-env NODE_ENV=production webpack",
+    "build:watch": "cross-env NODE_ENV=development webpack --watch",
+    "clean": "rimraf dist",
+    "start": "rimraf dist && cross-env NODE_ENV=development webpack --watch",
```
You could simplify this to `npm-run-all clean build:watch`

feeloor replied:
That makes sense, might be worth updating the same thing in the `execution-timeline` project as well (copied it from there).

andrewthauer commented on packages/schematics/src/app/utils.ts:
```diff
@@ -0,0 +1,207 @@
+import { getLatestNodeVersion, NpmRegistryPackage } from './npmjs-utils';
+import { Schema } from './schema';
+
+export function validateOptions(options: Schema): Rule {
```
Wondering if it would make sense to organize these into a `rules` directory rather then in a big `utils` file?

feeloor replied:
Maybe instead of a directory I could rename the files to be `rules.ts`?

andrewthauer commented on packages/schematics/webpack.config.js:
```diff
@@ -0,0 +1,76 @@
+const path = require('path');
```
NOTE: This is fine for now, but planning on consolidating/refactoring the entire repo to use webpack with a base config and allow for overrides per project.

andrewthauer commented on packages/schematics/webpack.config.js:
```diff
@@ -0,0 +1,76 @@
+  },
+
+  externals: {
+    'aws-sdk': 'aws-sdk',
```
Can we add a comment here describing this is necessary to workaround an issue with building.

feeloor replied:
Yeah, will do.

andrewthauer reviewed (changes requested):
Overall looks good. A few minor suggestions, comments. @egervari - Considering if you have lots of experience building schematics can you also review?

feeloor committed:
Addressing PR comments

andrewthauer commented:
@feeloor - One thing I forgot on the first pass. We should probably update the readme with instructions on how to use the schematics. Wondering if we move the manually setup into the /docs folder and just link to it in order to keep the file length down and more approachable. Thoughts?

feeloor commented:
> @feeloor - One thing I forgot on the first pass. We should probably update the readme with instructions on how to use the schematics. Wondering if we move the manually setup into the /docs folder and just link to it in order to keep the file length down and more approachable. Thoughts?

@andrewthauer Yeah I think it makes sense to not remove the manual setup in case someone has like a special case or want to see what's happening under the hood. 
I'll create a draft for the readme and put it up for review.

Commit:
update publish config

feeloor committed:
Adding readme for Schematics and updating root readme

feeloor committed:
Specifying files to publish to npm

feeloor committed:
Renaming utils file to rules

egervari commented on packages/schematics/src/app/rules.ts:
```diff
@@ -0,0 +1,207 @@
+  };
+}
+
+export function addPackageJsonDependencies(): Rule {
```
I think this code already exists in `import { addPackageJsonDependency, NodeDependency } from '@schematics/angular/utility/dependencies';` Any reason we did not use this? We can probably remove all this code in this method.

feeloor replied:
I'm using it under the hood. The reason why I wrap it is because I need to get the latest version. Angular has it "hard coded", I guess the update the file using their build scripts. But my function calls the `addPackageJsonDependency` once I got the version.

egervari commented on packages/schematics/src/app/rules.ts:
```diff
@@ -0,0 +1,207 @@
+
+export function validateOptions(options: Schema): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    if (!options.project || options.project === '') {
```
One nifty thing is that we don't have to throw an exception here. We can just inspect the `workspace.projects` object. In most cases, there should really only be 1 project there in that object, so you could find the key and then set the project name for them on the options object, thus saving the end-user some effort. This can be done before options is sent to other parts of the code. I don't think it's common to have more than 1 project honestly, and if they have more than 1, there's a good chance the first one is the one they want to instrument with augury.

IDEs should also pass in the project name. I know WebStorm does.

feeloor replied:
That's one thought I actually had as well, although I wasn't sure we wanted to just assume or install it to a project unless they specifically said yes to it. That's why I wanted to have the `Want to use DEFAULT_PROJECT (Y/n)?` to make sure we have the right project.
I do however agree with you about that probably the most common case there is only one project in the workspace.

I tried using the `"$source": "projectName"` as default value in the schematics, but then the prompt won't display. 

Thoughts? Do you think it's okay to default it to the first `workspace.projects` in case there is none entered?

egervari replied:
@feeloor I personally would default to workspace.projects. I think that's the intent of 99% of users honestly. No reason to bug them for information we can get automatically. If they don't like the default behaviour, they can always provide the project value manually. Many throw an error if 2 or more projects exist just to be safe, but even that might not be necessary.

feeloor replied:
@egervari - Yeah after looking at how `@angular/material` and other libraries does it I removed the prompt and use the `defaultProject`.

egervari commented on packages/schematics/src/app/rules.ts:
```diff
@@ -0,0 +1,207 @@
+    if (!scriptsNode) {
+      // Haven't found the script key, add it to the root of the package.json.
+      appendPropertyInAstObject(
+        recorder,
```
One thing I found really help was to put all of the ast code in its own package. It gets large/messy at times, and when committing to refactor it, can take up a lot of code on its own. I think it's best to separate concerns. Consider `src/app/ast` folder for this.

My general folder structure was something like

```
/ast (related operations in 1 file, used several files for different tasks)
/rules (one rule per file)
/schematics
/types (I put interfaces and functions that operate specifically on those interfaces here. Schema interface and related functions can go here, for example.)
```

Doing this also made things a lot more testable, especially testing ast functionality in isolation.

feeloor replied:
I like this structure, I'll refactor the code to follow this structure.

egervari commented on packages/schematics/src/app/rules.spec.ts:
```diff
@@ -0,0 +1,99 @@
+
+    const workspace = getWorkspace(tree);
+    const project = getProject(workspace, schemaOptions.project);
+    const projectTargets = getProjectTargets(project);
```
The above lines are used in 5 places. Can we refactor this?

egervari commented on packages/schematics/src/app/rules.ts:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    const sourceDir = normalize(project.sourceRoot);
+    const templSource = apply(url('./files'), [
```
`templateSource`?

egervari commented on README.md:
```diff
@@ -46,79 +46,28 @@ _NOTE: This tool is still experimental. Feedback is greatly appreciated :smile:_
 - [How to Use](docs/how-to-use.md)
 - [Architecture](docs/architecture.md)
 - [Development Guide](docs/development-guide.md)
+- [Manual Installation](docs/manual-installation.md)
```
I like how you opted to make the manual instructions harder to find to encourage the use of schematics. 👍

egervari commented on packages/schematics/src/app/npmjs-utils.ts:
```diff
@@ -0,0 +1,31 @@
+  version: string;
+}
+
+export function getLatestNodeVersion(packageName: string): Promise<NpmRegistryPackage> {
```
The import I showed you might also do this too.

feeloor replied:
I couldn't find this function, they seem to hard-code the latest Angular versions (I guess during their build process).

egervari commented on packages/schematics/src/app/rules.ts:
```diff
@@ -0,0 +1,207 @@
+
+    const sourceDir = normalize(project.sourceRoot);
+    const templSource = apply(url('./files'), [
+      template({
```
I think we can just pass `options` here without using spread.

egervari commented on packages/schematics/src/app/rules.ts:
```diff
@@ -0,0 +1,207 @@
+      !projectTargets.build.configurations ||
+      !projectTargets.serve ||
+      !projectTargets.serve.configurations
+    ) {
```
maybe refactor to `areProjectTargetsNotFound()`? When if statements get larger than 1 line, I tend to put them in methods/functions to simplify and document the intent of what the large expression actually means in human terms.

egervari reviewed (approved):
Sweet! Great job learning all of the schematics stuff. I know finding information, good samples, etc. is difficult, and you learned to incorporate a lot of different features in this one schematic. This should really improve adoption.

I added some comments on some refactorings you could make if you were inclined. I also indicated how to solve the lack of `options.project`, although as far as I remember, I think this was set automatically? Did they change it? Maybe it was WebStorm doing it for me. In any event, inspecting the workspace can get you the project name, prefix, etc.

Great work!

feeloor committed:
Refactoring schematics folder and file structure

feeloor committed:
Updating README

feeloor committed:
Removing prompt and use default project instead

feeloor commented:
@egervari @andrewthauer - I did a few changes: 
- I refactored the file + folder structure of the schematics as suggested.
- I updated the README.
- I removed the prompt of project name and used the default for now.

I'm satisfied with how it is right now and think we are ready for beta (?).

feeloor committed:
Removing code used to publish initial package

Commit:
Removed redundant config

Commit:
Updating webpack

Commit:
Update schematics version
